File: insights/src/event/event_grid_subscriber.py
import time
import threading
import queue
from flask import request, jsonify
import utils
from event.data_preprocessor import dict_to_product_candle, DataPreprocessor
from analytics.candle_technical_indicators import update_technical_indicators, update_technical_indicators_batch
from event.model.EventType import EventType

_logger = utils.get_logger("Insights")

BUFFER_SIZE_THRESHOLD = 10
DELAY_THRESHOLD = 5
WORKER_THREAD_COUNT = 6

# Shared in-memory structures
buffer_store = {}
buffer_lock = threading.Lock()
candle_event_queue = queue.Queue()
candle_store = {}  # NEW: shared across all threads

def handle_events():
    try:
        validation_response = validate_event_grid(request)
        if validation_response:
            return validation_response

        events = request.get_json()
        now = time.time()

        with buffer_lock:
            for event in events:
                event_id = event.get('id')
                event_type = event.get('eventType')
                if not event_id or not event_type:
                    _logger.error(f"Invalid event: {event}")
                    continue

                event['received_at'] = now

                if event_type == EventType.CANDLE:
                    buffer_store[event_id] = event
                    flushed = flush_buffer(now)
                    if flushed:
                        candle_event_queue.put(flushed)

                elif event_type == EventType.HISTORICAL_CANDLE:
                    _logger.info(f"Immediate processing: skipping buffer for batched event [Event I.D: {event_id}]")
                    threading.Thread(target=process_batched_event, args=(event,), daemon=True).start()

                else:
                    _logger.info(f"Unexpected event type [EventType: {event_type}, Event I.D: {event_id}]")

        return jsonify({"message": "Events handled."}), 200

    except Exception as e:
        _logger.exception(f"Exception handling event: {e}")
        return jsonify({"error": str(e)}), 400

def validate_event_grid(req):
    if req.headers.get('aeg-event-type') == 'SubscriptionValidation':
        validation_event = req.get_json()[0]
        validation_code = validation_event['data']['validationCode']
        return jsonify({'validationResponse': validation_code})
    return None

def flush_buffer(now):
    if not buffer_store:
        return None

    sorted_ids = sorted(buffer_store.keys())
    _logger.debug(f"Buffer size: {len(buffer_store)}. Sorted IDs: {sorted_ids}")

    for event_id in sorted_ids:
        event = buffer_store.get(event_id)
        if not event:
            continue
        if now - event['received_at'] >= DELAY_THRESHOLD:
            _logger.info(f"Flushing event [EventId: {event_id}] due to time delay")
            del buffer_store[event_id]
            return event

    if len(buffer_store) >= BUFFER_SIZE_THRESHOLD:
        oldest_id = sorted_ids[0]
        event = buffer_store.pop(oldest_id, None)
        if event:
            _logger.info(f"Flushing event [EventId: {oldest_id}] due to buffer size limit")
            return event

    return None

def process_ordered_event_worker():
    while True:
        event = candle_event_queue.get()
        try:
            data = event.get('data')
            event_id = event.get('id')
            _logger.info(f"[Worker] Processing CANDLE event [EventId: {event_id}]")
            product_candle = dict_to_product_candle(data)
            analysis = update_technical_indicators(product_candle, candle_store)
            DataPreprocessor().eventise_product_candle_analysis(event_id, analysis)
        except Exception as e:
            _logger.exception(f"[Worker] Failed to process CANDLE event: {e}")
        finally:
            candle_event_queue.task_done()

def process_batched_event(event):
    try:
        data = event.get('data')
        event_id = event.get('id')
        _logger.info(f"Processing {len(data)} HISTORICAL_CANDLE entries [EventId: {event_id}]")

        data_iterable = data if isinstance(data, list) else data.values()
        candles = [dict_to_product_candle(c) for c in data_iterable]
        batch = update_technical_indicators_batch(candles, candle_store)
        DataPreprocessor().eventise_product_candle_analysis_batch(event_id, batch)

    except Exception as e:
        _logger.exception(f"Exception in HISTORICAL_CANDLE event processing: {e}")

def start_candle_event_workers():
    for i in range(WORKER_THREAD_COUNT):
        thread = threading.Thread(target=process_ordered_event_worker, daemon=True)
        thread.start()
        _logger.info(f"Started worker thread #{i+1} for handling candle data events")

start_candle_event_workers()

File: core/src/server/event/event_grid_subscriber.js
const {getLogger, convertGmtToUKLocal} = require("../utils");
const {broadcastToClients} = require('../websocket/websocket_publisher');
const {insertDBAnalytics, insertDBAnalyticsBatch, readDBAnalytics} = require("../db/candle_analytics_cache");
const {EventType} = require("./model/EventType");
const { Mutex } = require('async-mutex');
const dbMutex = new Mutex();

const logger = getLogger();

let bufferStore = {};
const processedEventIds = new Set();
const DELAY_THRESHOLD = 2;
const BUFFER_SIZE_THRESHOLD = 5;

// === Batch processing queue + worker pool ===
const batchQueue = [];
const BATCH_WORKER_COUNT = 4;

function enqueueBatchEvent(event) {
    batchQueue.push(event);
}

async function batchWorker() {
    while (true) {
        if (batchQueue.length === 0) {
            await new Promise(resolve => setTimeout(resolve, 50)); // Wait 50ms if queue is empty
            continue;
        }

        const event = batchQueue.shift();
        try {
            const eventID = event.id;
            const receivedData = event.data;

            if (!Array.isArray(receivedData) || receivedData.length === 0) {
                logger.warn(`Skipped invalid or empty batch payload [EventID: ${eventID}]`);
                continue;
            }

            for (const candle of receivedData) {
                if (candle?.time) {
                    candle.time = convertGmtToUKLocal(candle.time);
                } else {
                    logger.warn(`Candle missing time field: ${JSON.stringify(candle)}`);
                }
            }

            logger.info(`BatchWorker processing batch size: ${receivedData.length} [EventID: ${eventID}]`);
            await dbMutex.runExclusive(() => {
                insertDBAnalyticsBatch(receivedData);
            });
            const cachedData = readDBAnalytics(receivedData[0]?.id, 0, receivedData[0]?.granularity_mins);
            broadcastToClients(cachedData, event.eventType, event.id);
        } catch (err) {
            logger.error(`BatchWorker failed to process batch event: ${err}`);
        }
    }
}

function startBatchWorkers() {
    for (let i = 0; i < BATCH_WORKER_COUNT; i++) {
        batchWorker(); // Start worker (no need to wrap in async IIFE)
        logger.info(`Started BatchWorker #${i + 1}`);
    }
}

// === Main Event Handler ===

function handleEvents(req, res) {
    try {
        const data = req.body;
        let validationResponse = validateEventGrid(req);
        if (validationResponse) {
            logger.info(`Responding to Event Grid validation request`);
            res.status(200).json(validationResponse);
            return;
        }

        processEvent(data);
        res.status(200).send({ success: true });
    } catch (error) {
        logger.error(`Error handling event: ${error}`);
        res.status(400).send({ success: false });
    }
}

function validateEventGrid(req) {
    if (req.headers['aeg-event-type'] === 'SubscriptionValidation') {
        let validationEvent = req.body[0];
        let validationCode = validationEvent.data.validationCode;
        return { validationResponse: validationCode };
    }
    return null;
}

function flushBuffer(now) {
    if (bufferStore) {
        const sortedIDs = Object.keys(bufferStore).sort();
        logger.debug(`Buffer size: ${sortedIDs.length}, Sorted IDs: ${sortedIDs}`);

        for (let eventID of sortedIDs) {
            const event = bufferStore[eventID];
            if (!event) continue;

            let elapsedSecondsInBuffer = Math.floor((now - event.receivedAt) / 1000);
            if (elapsedSecondsInBuffer >= DELAY_THRESHOLD) {
                logger.info(`Flushing event from buffer [Event I.D: ${eventID}] (time threshold met, waited ${elapsedSecondsInBuffer} seconds)`);
                delete bufferStore[eventID];
                return event;
            }
        }

        if (Object.keys(bufferStore).length >= BUFFER_SIZE_THRESHOLD) {
            let oldestEventID = sortedIDs[0];
            const event = bufferStore[oldestEventID];
            if (event) {
                logger.info(`Flushing event from buffer [Event I.D: ${oldestEventID}] (size threshold met, buffer size: ${Object.keys(bufferStore).length})`);
                delete bufferStore[oldestEventID];
                return event;
            }
        }
    }
}

async function processEvent(data) {
    for (const event of data) {
        const eventID = event.id;
        if (!eventID) {
            logger.warn("Attempted to process received event: EventID missing");
            continue;
        }
        const eventType = event.eventType;
        if (!eventType) {
            logger.warn("Attempted to process received event: EventType missing");
            continue;
        }

        if (processedEventIds.has(eventID)) {
            logger.warn(`Duplicate event received and skipped [EventId: ${eventID}]`);
            continue;
        }
        processedEventIds.add(eventID);

        logger.info(`Processing received event [EventType: ${eventType}, EventId: ${eventID}]`);
        const now = Date.now();
        event.receivedAt = now;

        let receivedData = event.data;
        let cachedData = {};

        if (eventType === EventType.CANDLE_ANALYTICS_BATCH) {
            logger.info(`Immediate processing: skipping buffer for batched data event [Event I.D: ${eventID}]`);
            logger.info(`Handling batch size: ${receivedData.length} [EventID: ${eventID}]`);
            enqueueBatchEvent(event); // Queue for workers to process
        } else if (eventType === EventType.CANDLE_ANALYTICS) {
            bufferStore[eventID] = event;
            const flushedEvent = flushBuffer(now);
            if (!flushedEvent) continue;

            receivedData = flushedEvent.data;
            if (!receivedData.id || !receivedData.time) {
                logger.warn(`Invalid candle object: missing id or time [EventId: ${eventID}]`);
                continue;
            }

            receivedData.time = convertGmtToUKLocal(receivedData.time);
            await dbMutex.runExclusive(() => {
                insertDBAnalytics(receivedData);
            });
            cachedData = readDBAnalytics(receivedData.id, 0, receivedData.granularity_mins);
            broadcastToClients(cachedData, flushedEvent.eventType, flushedEvent.id);
        } else {
            logger.warn(`Unexpected flushed data format: ${typeof receivedData} [EventId: ${eventID}]`);
            continue;
        }
    }
}

startBatchWorkers(); // Call this on module load

module.exports = { handleEvents };
